presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -398,6 +436,238 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+                    assignments,
+                    Optional.of(layoutHandle),
+                    TupleDomain.all(),
+                    TupleDomain.all());
+        }
+
+        private TableFinishNode createTemporaryTableWrite(
+                TableHandle tableHandle,
+                Map<Symbol, ColumnMetadata> temporaryTableColumns,
+                List<Symbol> outputs,
+                List<List<Symbol>> inputs,
+                List<PlanNode> sources,
+                Map<Symbol, Expression> constantExpressions,
+                PartitioningMetadata expectedPartitioningMetadata)
+        {
+            if (!constantExpressions.isEmpty()) {
+                List<Symbol> constantSymbols = ImmutableList.copyOf(constantExpressions.keySet());
+
+                // update outputs
+                outputs = ImmutableList.<Symbol>builder()
+                        .addAll(outputs)
+                        .addAll(constantSymbols)
+                        .build();
+
+                // update inputs
+                inputs = inputs.stream()
+                        .map(input -> ImmutableList.<Symbol>builder()
+                                .addAll(input)
+                                .addAll(constantSymbols)
+                                .build())
+                        .collect(toImmutableList());
+
+                // update sources
+                sources = sources.stream()
+                        .map(source -> {
+                            Assignments.Builder assignments = Assignments.builder();
+                            assignments.putIdentities(source.getOutputSymbols());
+                            constantSymbols.forEach(symbol -> assignments.put(symbol, constantExpressions.get(symbol)));
+                            return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
+                        })
+                        .collect(toImmutableList());
+            }
+
+            NewTableLayout insertLayout = metadata.getInsertLayout(session, tableHandle)
+                    // TODO: support insert into non partitioned table
+                    .orElseThrow(() -> new IllegalArgumentException("insertLayout for the temporary table must be present"));
+
+            PartitioningHandle partitioningHandle = expectedPartitioningMetadata.getPartitioningHandle();
+            List<String> partitionColumns = expectedPartitioningMetadata.getPartitionColumns();
+            ConnectorNewTableLayout expectedNewTableLayout = new ConnectorNewTableLayout(partitioningHandle.getConnectorHandle(), partitionColumns);
+            verify(insertLayout.getLayout().equals(expectedNewTableLayout), "unexpected new table layout");
+
+            Map<String, Symbol> columnNameToSymbol = temporaryTableColumns.entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getValue().getName(), Map.Entry::getKey));
+            List<Symbol> partitioningSymbols = partitionColumns.stream()
```

Comment:
these conversions between symbol and column name / metadata are so complicated :) . But I don't see a simple way here.

I think it's OK to keep it fo now, and at some point we might want to introduce a `MaterializedExchangeSymbolHandler` to hide all these complicated details.  ;)